Add a complexity scoring class for Metal and OpenGL

gw280 opened the pull request:
This adds an initial implementation of a complexity scoring class for Metal. Some notes:

* There are a lot of magic numbers in this file. These come from me processing benchmark data, and I have written long descriptions in the comments to give the reader an idea of where these numbers came from.
* drawTextBlob and drawVertices are incomplete right now, because we are unable to get the glyph count or the vertex count respectively. In their place, I've put in estimates where I can.
* The scores assigned are based off a baseline of 0.0005ms being a score of 100. This is a *very* rough estimate. With a 32-bit unsigned integer, this will allow us to score up to approximately 21 seconds before we overflow.
* Throughout the file I reference the constants `m` and `c`. This stems from `y=mx+c`, and details the line graph used to best fit the benchmark data for that particular usecase. *Important:* these constants are before dividing the benchmark data by the number of draw calls made.

https://github.com/flutter/flutter/issues/86728

Tests to come.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

zanderso commented on display_list/display_list_complexity_metal.cc:
```diff
@@ -0,0 +1,758 @@
+void DisplayListMetalComplexityCalculator::MetalHelper::AccumulateComplexity(
+    unsigned int complexity) {
+  // Check to see if we will overflow by accumulating this complexity score
+  if (std::numeric_limits<unsigned int>::max() - complexity < complexity) {
```
Consider allowing the upper bound to be customized.

zanderso commented on display_list/display_list_complexity_metal.cc:
```diff
@@ -0,0 +1,758 @@
+    return;
+  }
+  // The performance penalties seem fairly consistent percentage-wise
+  float non_hairline_penalty = 1.0f;
```
Right now this suggestion is a premature optimization, but just want to throw it out there in case it's useful later:

Consider avoiding floating point. You can do that by multiplying these factors by 10, and then dividing by 10 later when you need the final result.

Other strategies might be needed below, but we can delay thinking about them until they're needed.

zanderso commented on display_list/display_list_complexity_metal.cc:
```diff
@@ -0,0 +1,758 @@
+
+  // Calculate the impact of saveLayer
+  unsigned int save_layer_complexity;
+  if (save_layer_count_ > 200) {
```
Pulling these out to named constants with comments describing their effect would be helpful to future maintainers.

gw280 replied:
I think there are so many constants that it's best to just have a comment above each one. I'll move the comment out of `saveLayer` and into this method so that it's described at the place where it's actually used

gw280 commented on display_list/display_list_complexity_metal.h:
```diff
@@ -0,0 +1,154 @@
+                    SkScalar dpr) override;
+
+    // This method finalizes the complexity score calculation and returns it
+    unsigned int ComplexityScore();
```
I don't think we need it. INT_MAX for a uint32_t is about 4.2 billion, which represents about 21 seconds of wall clock time. If we have a single DisplayList that's exceeding 21 seconds to rasterise then we have a serious problem.

gw280 commented on display_list/display_list_complexity_metal.cc:
```diff
@@ -0,0 +1,758 @@
+  // Cost scales linearly with length approximately through the origin
+  // So a line length of 500 would roughly have 5x the complexity cost
+  // of a line length of 100
+  unsigned int complexity = distance * aa_penalty * non_hairline_penalty;
```
Apparently not

gw280 commented:
Big update:

* I've added a base class called DisplayListComplexityHelper where a lot of the common code between the two complexity calculators lives
* Added an OpenGL implementation now
* Fixed some minor bugs that were present before

Tests to come.

zanderso commented on display_list/display_list_complexity_helper.h:
```diff
@@ -0,0 +1,275 @@
+// b) Try and use integer arithmetic as much as possible.
+// c) If a specific draw op is logarithmic in complexity, do a best fit
+//    onto a linear equation within the range we expect to see the variables
+//    fall within.
```
Can you suggest an approach in this comment for adding a new DL op?

gw280 replied:
Done

zanderso reviewed (commented):
small nit: Comments should end with a '.'.

The license check is just asking for the new files to be listed in the golden file.

gw280 commented on flow/raster_cache.cc:
```diff
@@ -111,8 +111,8 @@ static bool IsDisplayListWorthRasterizing(
 
   // TODO(abarth): We should find a better heuristic here that lets us avoid
```
Done!
